images/base/files/usr/local/bin/entrypoint

```diff
@@ -372,6 +372,13 @@ enable_network_magic(){
       sed -i "s#${old_ipv4}#${curr_ipv4}#" /etc/kubernetes/manifests/*.yaml || true
       # this is no longer required with autodiscovery
       sed -i "s#${old_ipv4}#${curr_ipv4}#" /var/lib/kubelet/kubeadm-flags.env || true
+      # controller-manager and scheduler connect to local API endpoint. If the Node's IP changes, server address in
+      # kubeconfig should be updated. However, the server certificate isn't valid for the new IP. We update it to
+      # loopback address as it's an alternative address of the certificate.
+      if [[ "${old_ipv4}" != "${curr_ipv4}" ]]; then
+        sed -i "s#${old_ipv4}#127.0.0.1#" /etc/kubernetes/controller-manager.conf || true
```

Comment:
hmm,  @neolit123 how are these files generated?
can we influence them?

it seems easier to generate with the localhost IP instead of the node IP, they will only contact the local apiserver ... in HA we have one of these in each CP

Reply:
what I'm saying is to modify the code so the scheduler and the controller-manager always use API server ip address = `localhost` , maybe there is a knob or a flag to do it

Reply:
Thanks @aojea. I agree with you and I actually tried it but it seems more complicated.
The address in controller-manager.conf and scheduler.conf comes from `localAPIEndpoint.advertiseAddress`:
https://github.com/kubernetes-sigs/kind/blob/b4ab02cffbc8aca096ee7e2bc0f105654f1572c6/pkg/cluster/internal/kubeadm/config.go#L210-L211
https://github.com/kubernetes/kubernetes/blob/cdee77a4a9ba77a01bab6b462c1780d7a52af8c3/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go#L436

But `localAPIEndpoint.advertiseAddress` is used in many places in kubeadm so we can't simply replace it with localhost IP.

Reply:
the commit says https://github.com/kubernetes/kubernetes/commit/d9441906c4155173ce1a75421d8fcd1d2f79c471

```
Instead of using the CPE for these components, use the LocalAPIEndpoint.
This guarantees that the components would talk to the local
kube-apiserver, which should be the same version, unless the user
explicitly patched manifests.
```

wonder if we can patch it

Reply:
I guess it uses the local node IP instead of localhost because kubeadm doesn't add localhost to alternative addresses of the server certificate by default.
kind adds it for specific reason: https://github.com/kubernetes-sigs/kind/blob/b4ab02cffbc8aca096ee7e2bc0f105654f1572c6/pkg/cluster/internal/kubeadm/config.go#L167-L171.

Reply:
localAPIEndpoint.advertiseAddress is an IP because that's what kube-apiserver supports. Advertise / bind addresses for k8s components are IPs...they cannot be hostnames. So it must be 127.0.0.1 and not localhost. Just wanted to be clear about that.

I think I've seen users binding everything to loopback and it should work. I haven't tried it. If something in kubeadm is blocking you let me know. But also check if the manifest patches work for you ALA initconfiguration.patches